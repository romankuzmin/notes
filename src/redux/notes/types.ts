export type ProcessState = 'none' | 'fetching' | 'creating' | 'updating' | 'deleting' | 'rejected' | 'fulfilled';

export type Note = {
    id: string;
    title: string;
};

export type NoteState = {
    state: ProcessState;
    notify?: string;
    detail?: Note;
    list: Note[];
};
