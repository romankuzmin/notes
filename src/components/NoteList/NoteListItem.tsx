import { ListItem, ListItemText } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import React, { FC, forwardRef, memo, useMemo } from 'react';
import { Link, LinkProps as RouterLinkProps } from 'react-router-dom';

const useStyles = makeStyles({
    root: {
        listStyle: 'none',
    },
});

type NoteProps = {
    id: string;
    title: string;
};

const NoteListItem: FC<NoteProps> = ({ id, title }) => {
    const classes = useStyles();

    const CustomLink = useMemo(
        () =>
            // eslint-disable-next-line react/display-name
            forwardRef<never, Omit<RouterLinkProps, 'to'>>((linkProps, ref) => (
                <Link to={`/${id}`} {...linkProps} ref={ref} />
            )),
        [id],
    );

    return (
        <div className={classes.root}>
            <ListItem button={true} component={CustomLink}>
                <ListItemText primary={title} />
            </ListItem>
        </div>
    );
};

export default memo(NoteListItem);
