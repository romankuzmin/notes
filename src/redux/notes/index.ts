import { ActionCreatorWithPayload, AnyAction, createSlice, Draft, PayloadAction } from '@reduxjs/toolkit';
import { call, put, takeEvery } from 'redux-saga/effects';
import config from '../../config';
import { fetchApi } from '../../utils/fetch-api';
import { NoteState } from './types';

// First approach: define the initial state using that type
const initialState: NoteState = {
    state: 'none',
    list: [],
};

// @reduxjs/toolkit use Immer for immutability under the hood, so this code is immutable
const notesSlice = createSlice({
    name: 'notes',
    initialState,
    reducers: {
        reset: (state) => {
            state.state = 'none';
        },
        fetchNotes: (state) => {
            state.state = 'fetching';
        },
        fetchNotesFulfilled: (state, action) => {
            state.state = 'fulfilled';
            state.list = action.payload;
        },
        fetchNote: (state, action) => {
            state.state = 'fetching';
        },
        fetchNoteFulfilled: (state, action) => {
            state.state = 'fulfilled';
            state.detail = action.payload;
        },
        createNote: (state: Draft<NoteState>, action) => {
            state.state = 'creating';
        },
        createNoteFulFilled: (state, action) => {
            state.state = 'fulfilled';
            state.list.push(action.payload);
        },
        updateNote: (state, action) => {
            state.state = 'updating';
            state.notify = 'notes.detail.updating';
        },
        updateNoteFulfilled: (state, action) => {
            const { id, ...rest } = action.payload;

            state.state = 'fulfilled';
            if (state.detail) {
                state.detail.title = rest.title;
            }
        },
        deleteNote: (state: Draft<NoteState>, action: PayloadAction<unknown>) => {
            state.state = 'deleting';
            state.notify = 'notes.detail.deleting';
        },
        deleteNoteFulfilled: (state, action) => {
            const { id } = action.payload;
            const index = state.list.findIndex((item) => item.id === id);

            state.state = 'fulfilled';
            state.list.splice(index, 1);
        },
        error: (state, action) => {
            const error = action.payload;

            state.state = 'rejected';
            state.notify = error;
        },
        notify: (state, action) => {
            state.notify = action.payload;
            state.state = 'none';
        },
    },
});

// Extract the action creators object and the reducer
const { actions, reducer } = notesSlice;

const sagaActions = {
    FETCH_NOTES: 'notes/fetchNotes',
    FETCH_NOTE: 'notes/fetchNote',
    CREATE_NOTE: 'notes/createNote',
    UPDATE_NOTE: 'notes/updateNote',
    DELETE_NOTE: 'notes/deleteNote',
    NOTIFY_MESSAGE: 'notes/notifyMessage',
};

function* fetchNotesSaga() {
    yield processCall(fetchNotesFulfilled);
}

function* fetchNoteSaga(action: AnyAction) {
    const id = action.payload;

    yield processCall(fetchNoteFulfilled, id);
}

function* createNoteSaga(action: AnyAction) {
    yield processCall(createNoteFulFilled, undefined, 'POST', action.payload);
}

function* updateNoteSaga(action: AnyAction) {
    const { id, ...rest } = action.payload;
    yield processCall(updateNoteFulfilled, id, 'PUT', rest);
}

function* deleteNoteSaga(action: AnyAction) {
    const { id } = action.payload;
    yield processCall(deleteNoteFulfilled, id, 'DELETE');
}

function* processCall<T>(actionFulfilled: ActionCreatorWithPayload<T>, id?: string, method = 'GET', data?: T) {
    try {
        const apiUrl = `${config.apiUrl}${config.resource.notes}`;
        const url = id ? `${apiUrl}/${id}` : apiUrl;
        const result = yield call(() => fetchApi(url, method, data));
        yield put(actionFulfilled(result));
        yield put(reset());
    } catch (e) {
        console.error(e.message);
        let errorMessage =
            method === 'PUT' || method === 'POST' ? 'notes.detail.errors.updating' : 'notes.detail.errors.fetching';
        errorMessage = method === 'DELETE' ? 'notes.detail.errors.deleting' : errorMessage;

        yield put(error(errorMessage));
    }
}

export function* noteSaga() {
    yield takeEvery(sagaActions.FETCH_NOTES, fetchNotesSaga);
    yield takeEvery(sagaActions.FETCH_NOTE, fetchNoteSaga);
    yield takeEvery(sagaActions.CREATE_NOTE, createNoteSaga);
    yield takeEvery(sagaActions.UPDATE_NOTE, updateNoteSaga);
    yield takeEvery(sagaActions.DELETE_NOTE, deleteNoteSaga);
}

// Extract and export each action creator by name
export const {
    notify,
    error,
    reset,
    fetchNote,
    fetchNoteFulfilled,
    fetchNotes,
    fetchNotesFulfilled,
    createNote,
    createNoteFulFilled,
    updateNote,
    updateNoteFulfilled,
    deleteNote,
    deleteNoteFulfilled,
} = actions;

// Export the reducer, either as a default or named export
export default reducer;
