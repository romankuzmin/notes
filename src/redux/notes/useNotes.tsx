import { Draft, PayloadAction } from '@reduxjs/toolkit';
import { useEffect } from 'react';
import { useDispatch } from 'react-redux';
import { createNote, deleteNote, fetchNote, fetchNotes, notify, updateNote } from './index';
import { AppDispatch, useTypedSelector } from '../typed-hooks';
import { Note, ProcessState } from './types';

type NotesHook = {
    notes: Note[];
};

type NoteHook = {
    note: Note | undefined;
    createNote: (note: Note) => void;
    updateNote: (note: Note) => void;
    deleteNote: (note: Note) => void;
};

type WithProcessStateAndNotify<T> = T & {
    state: ProcessState;
    message: string | undefined;
    notifyMessage: (message: string) => void;
};

const dispatchAction = (dispatch: AppDispatch, action: (state: Draft<Note>) => PayloadAction<Note>) => (note: Note) => {
    dispatch(
        action({
            id: note.id,
            title: note.title,
        }),
    );
};

const notifyMessage = (dispatch: AppDispatch) => (message: string) => {
    dispatch(notify(message));
};

export const useNotes = () => {
    const dispatch = useDispatch<AppDispatch>();
    const { state, list, notify } = useTypedSelector((state) => state.notes);

    useEffect(() => {
        dispatch(fetchNotes());
    }, [dispatch]);

    return {
        notes: list,
        state: state,
        message: notify,
        notifyMessage: notifyMessage(dispatch),
    } as WithProcessStateAndNotify<NotesHook>;
};

export const useNote = (id?: string) => {
    const dispatch = useDispatch<AppDispatch>();
    const { state, detail, notify } = useTypedSelector((state) => state.notes);

    useEffect(() => {
        if (id) {
            dispatch(fetchNote(id));
        }
    }, [dispatch, id]);

    return {
        note: id ? detail : undefined,
        state: state,
        message: notify,
        createNote: dispatchAction(dispatch, createNote),
        updateNote: dispatchAction(dispatch, updateNote),
        deleteNote: dispatchAction(dispatch, deleteNote),
        notifyMessage: notifyMessage(dispatch),
    } as WithProcessStateAndNotify<NoteHook>;
};
