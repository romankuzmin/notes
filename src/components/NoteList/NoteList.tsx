import { Hidden, List, ListItem, ListItemText, Paper } from '@material-ui/core';
import { makeStyles } from '@material-ui/core/styles';
import Skeleton from '@material-ui/lab/Skeleton';
import React, { FC, memo } from 'react';
import { FormattedMessage } from 'react-intl';
import NoteListItem from './NoteListItem';

const useStyles = makeStyles({
    root: {
        // maxWidth: '800px',
        // margin: 'auto 0',
    },
});

type NoteProps = {
    id: string;
    title: string;
};

type NoteListProps = {
    items: NoteProps[];
    loading?: boolean;
};

const generateLoadingList = (length = 10) => {
    return new Array(length).fill('').map((item, index) => ({ id: `${index + 1}`, title: '' } as NoteProps));
};

const NoteList: FC<NoteListProps> = ({ items = [], loading = false }) => {
    const classes = useStyles();
    const list = loading ? generateLoadingList() : items;

    const noteItems = (
        <List component="nav">
            {list.map((item) =>
                loading ? (
                    <ListItem key={item.id}>
                        <ListItemText primary={<Skeleton animation="wave" variant="text" height={15} />} />
                    </ListItem>
                ) : (
                    <NoteListItem key={item.id} {...item} />
                ),
            )}
            {list.length === 0 && !loading && <ListItemText primary={<FormattedMessage id={'notes.empty'} />} />}
        </List>
    );

    return (
        <div className={classes.root}>
            <Hidden xsDown>
                <Paper>{noteItems}</Paper>
            </Hidden>
            <Hidden smUp>{noteItems}</Hidden>
        </div>
    );
};

export default memo(NoteList);
