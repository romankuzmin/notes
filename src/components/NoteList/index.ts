import NoteList from './NoteList';
import NoteListItem from './NoteListItem';

export { NoteList, NoteListItem };
