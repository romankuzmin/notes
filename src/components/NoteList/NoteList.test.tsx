import '@testing-library/jest-dom/extend-expect';
import { render } from '@testing-library/react';
import mediaQuery from 'css-mediaquery';
import React from 'react';
import { IntlProvider } from 'react-intl';
import { Note } from '../../redux/notes/types';

import NoteList from './NoteList';

function createMatchMedia(width: number) {
    return (query: string) =>
        (({
            matches: mediaQuery.match(query, { width }),
            addListener: () => {},
            removeListener: () => {},
        } as unknown) as MediaQueryList);
}

describe('NoteList', () => {
    beforeAll(() => {
        window.matchMedia = createMatchMedia(window.innerWidth);
    });

    test('Render blank list', () => {
        const { asFragment } = render(
            <IntlProvider locale="en" messages={{ 'notes.list.blank': "Notes are blank. It's time to create one ..." }}>
                <NoteList items={[]} />,
            </IntlProvider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('Render list with loading state', () => {
        const { asFragment } = render(
            <IntlProvider locale="en" messages={{}}>
                <NoteList items={[]} loading={true} />,
            </IntlProvider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });

    test('Render list with two items', () => {
        const items: Note[] = [
            {
                id: '1',
                title: 'Note no. 1',
            },
            {
                id: '2',
                title: 'Note no. 2',
            },
        ];

        const { asFragment } = render(
            <IntlProvider locale="en" messages={{}}>
                <NoteList items={items} loading={true} />,
            </IntlProvider>,
        );

        expect(asFragment()).toMatchSnapshot();
    });
});
