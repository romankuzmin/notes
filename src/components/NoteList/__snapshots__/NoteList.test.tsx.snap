// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NoteList Render blank list 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <nav
      class="MuiList-root MuiList-padding"
    >
      <div
        class="MuiListItemText-root makeStyles-emptyList-1"
      >
        <span
          class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
        >
          Notes are blank. It's time to create one ...
        </span>
      </div>
    </nav>
  </div>
  ,
</DocumentFragment>
`;

exports[`NoteList Render list with loading state 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <nav
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
    </nav>
  </div>
  ,
</DocumentFragment>
`;

exports[`NoteList Render list with two items 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <nav
      class="MuiList-root MuiList-padding"
    >
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
      <li
        class="MuiListItem-root MuiListItem-gutters"
      >
        <div
          class="MuiListItemText-root"
        >
          <span
            class="MuiTypography-root MuiListItemText-primary MuiTypography-body1 MuiTypography-displayBlock"
          >
            <span
              class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-wave"
              style="height: 15px;"
            />
          </span>
        </div>
      </li>
    </nav>
  </div>
  ,
</DocumentFragment>
`;
